app/retail/emails.py

```diff
@@ -1622,6 +1623,117 @@ def roundup(request):
     return HttpResponse(response_html)
 
 
+def daily_digest(request):
+    email = request.user.email if request.user.is_authenticated else 'test@123.com'
+    response_html, _, _, _, _ = render_daily_digest(email)
+    return HttpResponse(response_html)
+
+
+
+def hackathon_marketing(request):
+    email = request.user.email if request.user.is_authenticated else 'test@123.com'
+    response_html, _, _, _, _ = render_hackathon_marketing_email(email)
+    return HttpResponse(response_html)
+
+
+def render_daily_digest(to_email):
+    from dashboard.models import Bounty
+    from django.conf import settings
+    from marketing.models import DailyDigestEmail
+    args = DailyDigestEmail.objects.order_by('created_on').last()
+
+    subject = args.subject
+    if settings.DEBUG and False:
+        # for debugging email styles
+        email_style = 2
+    else:
+        offset = 2
+        email_style = (int(timezone.now().strftime("%V")) + offset) % 7
+
+    intro = args.body
+    bounties_spec = args.bounties_spec
```

Comment:
L1654 - L1668 is identical to L1701 - L1715

you can use a function

```
def get_bounties(spec):
    bounties = []
    for nb in spec:
        try:
            bounty = Bounty.objects.current().filter(
                github_url__iexact=nb['url'],
            ).order_by('-web3_created').first()
            if bounty:
                bounties.append({
                    'obj': bounty,
                    'primer': nb['primer']
                })
        except Exception as e:
            print(e)
    return bounties
```

and call `get_bounties(args.bounties_spec)`